In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import os
os.getcwd()

'C:\\Users\\prati'

In [3]:

data = pd.read_csv(r"C:\Users\prati\Downloads\archive (9)\ifood_df.csv")
data

Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  \
0     58138.0        0         0       58       635         88   
1     46344.0        1         1       38        11          1   
2     71613.0        0         0       26       426         49   
3     26646.0        1         0       26        11          4   
4     58293.0        1         0       94       173         43   
...       ...      ...       ...      ...       ...        ...   
2200  61223.0        0         1       46       709         43   
2201  64014.0        2         1       56       406          0   
2202  56981.0        0         0       91       908         48   
2203  69245.0        0         1        8       428         30   
2204  52869.0        1         1       40        84          3   

      MntMeatProducts  MntFishProducts  MntSweetProducts  MntGoldProds  ...  \
0                 546              172                88            88  ...   
1                   6                2                 1             6  ...   
2                 127              111                21            42  ...   
3                  20               10                 3             5  ...   
4                 118               46                27            15  ...   
...               ...              ...               ...           ...  ...   
2200              182               42               118           247  ...   
2201               30                0                 0             8  ...   
2202              217               32                12            24  ...   
2203              214               80                30            61  ...   
2204               61                2                 1            21  ...   

      marital_Together  marital_Widow  education_2n Cycle  education_Basic  \
0                    0              0                   0                0   
1                    0              0                   0                0   
2                    1              0                   0                0   
3                    1              0                   0                0   
4                    0              0                   0                0   
...                ...            ...                 ...              ...   
2200                 0              0                   0                0   
2201                 1              0                   0                0   
2202                 0              0                   0                0   
2203                 1              0                   0                0   
2204                 0              0                   0                0   

      education_Graduation  education_Master  education_PhD  MntTotal  \
0                        1                 0              0      1529   
1                        1                 0              0        21   
2                        1                 0              0       734   
3                        1                 0              0        48   
4                        0                 0              1       407   
...                    ...               ...            ...       ...   
2200                     1                 0              0      1094   
2201                     0                 0              1       436   
2202                     1                 0              0      1217   
2203                     0                 1              0       782   
2204                     0                 0              1       151   

      MntRegularProds  AcceptedCmpOverall  
0                1441                   0  
1                  15                   0  
2                 692                   0  
3                  43                   0  
4                 392                   0  
...               ...                 ...  
2200              847                   0  
2201              428                   1  
2202             1193                   1  
2203       

In [4]:
data.head(5)

Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  MntMeatProducts  \
0  58138.0        0         0       58       635         88              546   
1  46344.0        1         1       38        11          1                6   
2  71613.0        0         0       26       426         49              127   
3  26646.0        1         0       26        11          4               20   
4  58293.0        1         0       94       173         43              118   

   MntFishProducts  MntSweetProducts  MntGoldProds  ...  marital_Together  \
0              172                88            88  ...                 0   
1                2                 1             6  ...                 0   
2              111                21            42  ...                 1   
3               10                 3             5  ...                 1   
4               46                27            15  ...                 0   

   marital_Widow  education_2n Cycle  education_Basic  education_Graduation  \
0              0                   0                0                     1   
1              0                   0                0                     1   
2              0                   0                0                     1   
3              0                   0                0                     1   
4              0                   0                0                     0   

   education_Master  education_PhD  MntTotal  MntRegularProds  \
0                 0              0      1529             1441   
1                 0              0        21               15   
2                 0              0       734              692   
3                 0              0        48               43   
4                 0              1       407              392   

   AcceptedCmpOverall  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 39 columns]

In [5]:
data.shape

(2205, 39)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2205 entries, 0 to 2204
Data columns (total 39 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Income                2205 non-null   float64
 1   Kidhome               2205 non-null   int64  
 2   Teenhome              2205 non-null   int64  
 3   Recency               2205 non-null   int64  
 4   MntWines              2205 non-null   int64  
 5   MntFruits             2205 non-null   int64  
 6   MntMeatProducts       2205 non-null   int64  
 7   MntFishProducts       2205 non-null   int64  
 8   MntSweetProducts      2205 non-null   int64  
 9   MntGoldProds          2205 non-null   int64  
 10  NumDealsPurchases     2205 non-null   int64  
 11  NumWebPurchases       2205 non-null   int64  
 12  NumCatalogPurchases   2205 non-null   int64  
 13  NumStorePurchases     2205 non-null   int64  
 14  NumWebVisitsMonth     2205 non-null   int64  
 15  AcceptedCmp3         

In [7]:
data.describe()

Income      Kidhome     Teenhome      Recency     MntWines  \
count    2205.000000  2205.000000  2205.000000  2205.000000  2205.000000   
mean    51622.094785     0.442177     0.506576    49.009070   306.164626   
std     20713.063826     0.537132     0.544380    28.932111   337.493839   
min      1730.000000     0.000000     0.000000     0.000000     0.000000   
25%     35196.000000     0.000000     0.000000    24.000000    24.000000   
50%     51287.000000     0.000000     0.000000    49.000000   178.000000   
75%     68281.000000     1.000000     1.000000    74.000000   507.000000   
max    113734.000000     2.000000     2.000000    99.000000  1493.000000   

         MntFruits  MntMeatProducts  MntFishProducts  MntSweetProducts  \
count  2205.000000      2205.000000      2205.000000       2205.000000   
mean     26.403175       165.312018        37.756463         27.128345   
std      39.784484       217.784507        54.824635         41.130468   
min       0.000000         0.000000         0.000000          0.000000   
25%       2.000000        16.000000         3.000000          1.000000   
50%       8.000000        68.000000        12.000000          8.000000   
75%      33.000000       232.000000        50.000000         34.000000   
max     199.000000      1725.000000       259.000000        262.000000   

       MntGoldProds  ...  marital_Together  marital_Widow  education_2n Cycle  \
count   2205.000000  ...       2205.000000    2205.000000         2205.000000   
mean      44.057143  ...          0.257596       0.034467            0.089796   
std       51.736211  ...          0.437410       0.182467            0.285954   
min        0.000000  ...          0.000000       0.000000            0.000000   
25%        9.000000  ...          0.000000       0.000000            0.000000   
50%       25.000000  ...          0.000000       0.000000            0.000000   
75%       56.000000  ...          1.000000       0.000000            0.000000   
max      321.000000  ...          1.000000       1.000000            1.000000   

       education_Basic  education_Graduation  education_Master  education_PhD  \
count      2205.000000           2205.000000       2205.000000    2205.000000   
mean          0.024490              0.504762          0.165079       0.215873   
std           0.154599              0.500091          0.371336       0.411520   
min           0.000000              0.000000          0.000000       0.000000   
25%           0.000000              0.000000          0.000000       0.000000   
50%           0.000000              1.000000          0.000000       0.000000   
75%           0.000000              1.000000          0.000000       0.000000   
max           1.000000              1.000000          1.000000       1.000000   

          MntTotal  MntRegularProds  AcceptedCmpOverall  
count  2205.000000      2205.000000          2205.00000  
mean    562.764626       518.707483             0.29932  
std     575.936911       553.847248             0.68044  
min       4.000000      -283.000000             0.00000  
25%      56.000000        42.000000             0.00000  
50%     343.000000       288.000000             0.00000  
75%     964.000000       884.000000             0.00000  
max    2491.000000      2458.000000             4.00000  

[8 rows x 39 columns]

In [8]:
# Checking for null values

data.isnull().sum()

Income                  0
Kidhome                 0
Teenhome                0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
Age                     0
Customer_Days           0
marital_Divorced        0
marital_Married         0
marital_Single          0
marital_Together        0
marital_Widow           0
education_2n Cycle      0
education_Basic         0
education_Graduation    0
education_Master        0
education_PhD           0
MntTotal                0
MntRegularProds         0
AcceptedCmpO

In [9]:
'''Create a Working Copy

Never clean on raw data.'''

df = data.copy()

In [10]:
df["Z_CostContact"]

0       3
1       3
2       3
3       3
4       3
       ..
2200    3
2201    3
2202    3
2203    3
2204    3
Name: Z_CostContact, Length: 2205, dtype: int64

In [11]:
# Deleting constant value columns as they add no value
# Constants add no value.

data.drop(columns = {"Z_CostContact", "Z_Revenue"},inplace=True)

In [12]:
data.columns

Index(['Income', 'Kidhome', 'Teenhome', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Response', 'Age', 'Customer_Days',
       'marital_Divorced', 'marital_Married', 'marital_Single',
       'marital_Together', 'marital_Widow', 'education_2n Cycle',
       'education_Basic', 'education_Graduation', 'education_Master',
       'education_PhD', 'MntTotal', 'MntRegularProds', 'AcceptedCmpOverall'],
      dtype='object')

In [13]:
# Dropped duplicates

data[data.duplicated()]
data.drop_duplicates(inplace=True)

In [14]:
data.shape

(2021, 37)

In [18]:
'''
Frequency → shows how often customers buy (behavior)
Tenure → separates new vs long-term customers

These two are mandatory for:
baseline understanding
responder vs non-responder analysis
segmentation later'''

# Purchase frequency
data["Frequency"] = (data["NumWebPurchases"] +data["NumCatalogPurchases"] +data["NumStorePurchases"])

# Customer tenure in years
data["Tenure_Years"] = data["Customer_Days"] / 365

data["Frequency"]
data["Tenure_Years"]

0       7.731507
1       6.224658
2       6.769863
3       6.295890
4       6.356164
          ...   
2198    7.780822
2200    6.958904
2202    6.339726
2203    6.342466
2204    7.619178
Name: Tenure_Years, Length: 2021, dtype: float64

In [25]:
'''Establish the baseline
Creates the reference point for all comparisons
Prevents misleading insights later
This is how analysts stay objective'''

# Total customers
total_customers = data.shape[0]

# Average & median spend
avg_spend = data["MntTotal"].mean()
median_spend = data["MntTotal"].median()

# Average income
avg_income = data["Income"].mean()

# Average purchase frequency
avg_frequency = data["Frequency"].mean()

# Campaign response rate
response_rate = data["Response"].mean()

total_customers, avg_spend, median_spend, avg_income, avg_frequency, response_rate


(2021,
 np.float64(563.789213260762),
 np.float64(343.0),
 np.float64(51687.2587827808),
 np.float64(12.567540821375557),
 np.float64(0.15388421573478475))

In [26]:
data["MntTotal"]

0       1529
1         21
2        734
3         48
4        407
        ... 
2198      19
2200    1094
2202    1217
2203     782
2204     151
Name: MntTotal, Length: 2021, dtype: int64

In [27]:
'''Revenue concentration (who drives the money)
Identify how much revenue comes from the top customers.
Shows whether revenue is concentrated or spread
Helps marketing focus on high-impact customers
This is a classic executive-level insight'''

# Sort customers by total spend
data_sorted = data.sort_values("MntTotal", ascending=False)

# Top 20% customers
top_20_pct = int(0.2 * len(data_sorted))
top_customers = data_sorted.head(top_20_pct)

# Revenue contribution
total_revenue = data_sorted["MntTotal"].sum()
top_revenue = top_customers["MntTotal"].sum()

top_revenue_share = top_revenue / total_revenue
top_revenue_share


np.float64(0.5357068257654346)

Top 20% customers generate 53.6% of the total revenue.
Losing top customers would hurt revenue significantly.
Marketing should prioritize retention and upsell for this group

In [29]:
'''Responder vs Non-Responder comparison
Split customers based on campaign response and compare core metrics.
Identifies who actually responds to marketing
Directly informs targeting strategy
This is the core client question'''

responders = data[data["Response"]==1]
non_responders = data[data["Response"] == 0]

comparison = {
    "Avg_Income": [
        responders["Income"].mean(),
        non_responders["Income"].mean()
    ],
    "Avg_Total_Spend": [
        responders["MntTotal"].mean(),
        non_responders["MntTotal"].mean()
    ],
    "Avg_Frequency": [
        responders["Frequency"].mean(),
        non_responders["Frequency"].mean()
    ],
    "Avg_Recency": [
        responders["Recency"].mean(),
        non_responders["Recency"].mean()
    ]
}

comparison

{'Avg_Income': [np.float64(60192.906752411574), np.float64(50140.32514619883)],
 'Avg_Total_Spend': [np.float64(917.2636655948553),
  np.float64(499.5023391812866)],
 'Avg_Frequency': [np.float64(15.270096463022508),
  np.float64(12.076023391812866)],
 'Avg_Recency': [np.float64(35.038585209003216),
  np.float64(51.39824561403509)]}

In [36]:
'''Product-level revenue contribution
Which product categories drive revenue.
Tells what to promote in campaigns
Prevents wasting effort on low-impact products
Directly informs offer design'''

product_cols = ["MntWines", "MntFruits", "MntMeatProducts", "MntFishProducts", "MntSweetProducts", "MntGoldProds"]

product_revenue = data[product_cols].sum()
product_revenue

product_share = product_revenue / product_revenue.sum()
product_share

product_share.sort_values(ascending=False)



MntWines            0.504339
MntMeatProducts     0.273255
MntGoldProds        0.072274
MntFishProducts     0.061878
MntSweetProducts    0.044871
MntFruits           0.043384
dtype: float64

In [38]:
'''
Channel preference analysis
Determine how customers prefer to purchase.
Tells where to run campaigns
Prevents channel misalignment
Needed for Power BI storytelling'''


data["Web_Share"] = data["NumWebPurchases"] / data["Frequency"]
data["Store_Share"] = data["NumStorePurchases"] / data["Frequency"]
data["Catalog_Share"] = data["NumCatalogPurchases"] / data["Frequency"]

channel_pref = data[["Web_Share", "Store_Share", "Catalog_Share"]].mean()
channel_pref


Web_Share        0.331565
Store_Share      0.504097
Catalog_Share    0.164338
dtype: float64

In [39]:
'''
Simple customer segmentation
Create value-based segments using total spend
Marketing works with segments, not raw customers
This enables targeting, exclusion, and prioritization
Required for Power BI dashboards'''

# Create spend quantiles
high_value_threshold = data["MntTotal"].quantile(0.75)
low_value_threshold = data["MntTotal"].quantile(0.25)

def value_segment(spend):
    if spend >= high_value_threshold:
        return "High Value"
    elif spend <= low_value_threshold:
        return "Low Value"
    else:
        return "Mid Value"

data["Value_Segment"] = data["MntTotal"].apply(value_segment)

# Check segment distribution
data["Value_Segment"].value_counts(normalize=True)


Value_Segment
Mid Value     0.499258
High Value    0.250371
Low Value     0.250371
Name: proportion, dtype: float64